### ISSUE TITLE
[Bug] In rare cases the getblock RPC call incorrectly returns height: null and confirmations: 0

### ISSUE THREAD

**Issue by e-nikolov** [AleoHQ/snarkOS#900]
## 🐛 Bug Report

While working on the aleo explorer, I noticed that once every couple of hours while the explorer is fetching blocks from the "http://128.199.15.82:3030" boot node, it responds with something like this:
```
 {
   confirmations: 0,
   difficulty_target: 1844674407370955300,
   hash: '9b395f3efb734c5c4a7f0145be92a0f717cc37e24d4185997de0bde5a27e44f4',
   height: null,
   merkle_root: '79d29086bce735d5f5e84e86856e2bf73b9c863518f405db05604d8ea41a94df',
   nonce: 4279091154,
   pedersen_merkle_root_hash: '04a019da3a62a1c107600b52fa32fc04abe08532e4b5f226afe344359eabbb00',
   previous_block_hash: '4c13c0e7e8d9226656a36b79c55da9fda67d30bc982cf192fd773ecb53bc6409',
   proof: '030000000000000004000000000000004c2c8933707ca3cb151ac85711ee382005ec518ddde78f5d379e456cce0740e10b4c2cae73fcf426698630f7f745f6000025ca9597f9d70358cf236ecebeda2ecb3e892994f45776320b63c40eecddbefd07b82f0ec4bf087f5f91e6b00684220000d7f69ccc24ef41fa114a4998f4a6252c7e90140c2845dadf64f647e92baa520978c111822b27442c335271ac0ea099000017169ac83c420003bcd072445ae19f4fbb85351def879f23213689679e9d2f9c93016531d3c1816209a5126ce669ed8000030000000000000028b8c1b9a42d6e33f9de2b2f6d7142d95e9ad21eb06760a29868eba7832c25188e5a089538e8d5363da1f1b373073f80005ff15c3bb5953ab67546b0bbf20eae2ff2d6c0262827ed85bd63286613298396c1fb21f0bb7dd37a61683e386933798001ae9bf321785b7e4cd16ed53eefb82e9d3ace7a287c1fa19b42203d13b9fdf813322a993bb70c0229c7fef81647d1d300dacf5f4920a3c0187617cecee839e5986416c325966b25002c3932912a703cf608837e5681ca248029f9ca7487ed8280000200000000000000a6e2d0002e1386e484757cd351fd638bb74faf30426e8771f67a9e52f9d27dbaf2297d0cd25e2d46fc4197905510940001dd35cbd0b2c89def598420564686971a9d8dd3ad1cd921080334104f6c25da5d7b29891d3ff78ac15fb7d88ee6f6ab80cf3cdb7b9e7fb6b8599b3014c4023b3efef30e343451947def094b1f95bf259b1df7ee3879af30219c15bd063a750f8100070000000000000013873c637bbeef85241db8df4554f1d175ea047e69f83f78e1d6323bb916130ae6ab73d198c975157d712ad758ec124b1ad3c5d065609a93aa153e033d60a30dbbc46d66c10251ea4afd810e711c4607f2c9a9b2bb0f3b2eb14094fdec41b9013b5c7f2f118243160c22879eb86cf2eb79a3237baa5f8c6072c171f9d00aa911efc70c56a26c600f57906e057420f361566425694c841d3b3088065d6b87ea02c17f6c0e66e1e8705f604d010c214bb9edb1f7add514e2ad9f92fc2595109310b6d68726a59f0c285b949296e19c3d3078b19f04ac2ca38d3b919dc19caeaf09030000000000000000000000000000000000000000000000000000000000000002000000000000000a9b77d86f3e87071776b26e0c46280243f62cb671ee3c5622b6293616857d3a249626f471a96f9c57fdf663178043800104fa182586e4f6a0816fcff0439d41718e133350ecf8b9ca3d159f80b400bc0751083c882aa402122e23686b685ecfaf361f65abc97a1fa5da6c87e8b54bca8eef4ef478db448189df4749ffa8e3f9800000',
   size: 2627,
   time: 1606562661,
   transactions: [
     '79d29086bce735d5f5e84e86856e2bf73b9c863518f405db05604d8ea41a94df'
   ]
 }
```

Notice that the height is null and confirmations is 0. But if I retry the same request manually, I get the correct values:
```
 {
   "confirmations":208776,
   "difficulty_target":1844674407370955161,
   "hash":"9b395f3efb734c5c4a7f0145be92a0f717cc37e24d4185997de0bde5a27e44f4",
   "height":93638,
   "merkle_root":"79d29086bce735d5f5e84e86856e2bf73b9c863518f405db05604d8ea41a94df",
   "nonce":4279091154,
   "pedersen_merkle_root_hash":"04a019da3a62a1c107600b52fa32fc04abe08532e4b5f226afe344359eabbb00",
   "previous_block_hash":"4c13c0e7e8d9226656a36b79c55da9fda67d30bc982cf192fd773ecb53bc6409",
   "proof":"030000000000000004000000000000004c2c8933707ca3cb151ac85711ee382005ec518ddde78f5d379e456cce0740e10b4c2cae73fcf426698630f7f745f6000025ca9597f9d70358cf236ecebeda2ecb3e892994f45776320b63c40eecddbefd07b82f0ec4bf087f5f91e6b00684220000d7f69ccc24ef41fa114a4998f4a6252c7e90140c2845dadf64f647e92baa520978c111822b27442c335271ac0ea099000017169ac83c420003bcd072445ae19f4fbb85351def879f23213689679e9d2f9c93016531d3c1816209a5126ce669ed8000030000000000000028b8c1b9a42d6e33f9de2b2f6d7142d95e9ad21eb06760a29868eba7832c25188e5a089538e8d5363da1f1b373073f80005ff15c3bb5953ab67546b0bbf20eae2ff2d6c0262827ed85bd63286613298396c1fb21f0bb7dd37a61683e386933798001ae9bf321785b7e4cd16ed53eefb82e9d3ace7a287c1fa19b42203d13b9fdf813322a993bb70c0229c7fef81647d1d300dacf5f4920a3c0187617cecee839e5986416c325966b25002c3932912a703cf608837e5681ca248029f9ca7487ed8280000200000000000000a6e2d0002e1386e484757cd351fd638bb74faf30426e8771f67a9e52f9d27dbaf2297d0cd25e2d46fc4197905510940001dd35cbd0b2c89def598420564686971a9d8dd3ad1cd921080334104f6c25da5d7b29891d3ff78ac15fb7d88ee6f6ab80cf3cdb7b9e7fb6b8599b3014c4023b3efef30e343451947def094b1f95bf259b1df7ee3879af30219c15bd063a750f8100070000000000000013873c637bbeef85241db8df4554f1d175ea047e69f83f78e1d6323bb916130ae6ab73d198c975157d712ad758ec124b1ad3c5d065609a93aa153e033d60a30dbbc46d66c10251ea4afd810e711c4607f2c9a9b2bb0f3b2eb14094fdec41b9013b5c7f2f118243160c22879eb86cf2eb79a3237baa5f8c6072c171f9d00aa911efc70c56a26c600f57906e057420f361566425694c841d3b3088065d6b87ea02c17f6c0e66e1e8705f604d010c214bb9edb1f7add514e2ad9f92fc2595109310b6d68726a59f0c285b949296e19c3d3078b19f04ac2ca38d3b919dc19caeaf09030000000000000000000000000000000000000000000000000000000000000002000000000000000a9b77d86f3e87071776b26e0c46280243f62cb671ee3c5622b6293616857d3a249626f471a96f9c57fdf663178043800104fa182586e4f6a0816fcff0439d41718e133350ecf8b9ca3d159f80b400bc0751083c882aa402122e23686b685ecfaf361f65abc97a1fa5da6c87e8b54bca8eef4ef478db448189df4749ffa8e3f9800000",
   "size":2627,
   "time":1606562661,
   "transactions":["79d29086bce735d5f5e84e86856e2bf73b9c863518f405db05604d8ea41a94df"]
 }
```

This is a fairly old block so it is unlikely that the node is still trying to decide if it's a fork or not.
After noticing this I wrote a script that retries getting the same block over and over and indeed once an hour or so it also returns height: null instead of the usual value.

I also noticed that usually when this issue happens, it affects all blocks fetched within that second.

For the explorer, this issue is most noticeable while the explorer is syncing from the beginning because at that point it fetches a lot of blocks within a short period of time.

**Comment by ljedrz** [comment]
Noteworthy: the `difficulty_target` is different between the 2 values returned by the RPC.

**Comment by e-nikolov** [comment]
I noticed that when the issue occurs, it seems to almost always return that same `1844674407370955300` difficulty target.

> How many blocks are you fetching within a second?

I tried different values in my script, between 100 and 10 000 requests per second, and I haven't run into rate limits as far as I can tell. Increasing the number of requests per second above a certain threshold doesn't seem to significantly increase the likelihood of getting the issue. I have a suspicion that the issue happens when the node is in some sort of a broken state that lasts less than a second and happens once in a while. Increasing the number of requests per second only increases the probability of having a request in flight during the node's broken state but it doesn't seem to cause it.

Is it possible that it happens when the node is starting up or during a database compaction at which point it can't access the latest storage?

**Comment by ljedrz** [comment]
Could you share the script you're using? I'd like to verify my secondary storage instance suspicions by testing it locally on a branch that doesn't use it; alternatively, you can take stab at trying to reproduce it on [that branch](https://github.com/ljedrz/snarkOS/tree/no_secondary_storage_for_rpc) yourself if you prefer (note: that branch also returns `null` for `confirmations` instead of `0` in case the block number can't be obtained).

**Comment by ljedrz** [comment]
Note: I'm getting `"difficulty_target": 1844674407370955300` with a valid height too.

**Comment by e-nikolov** [comment]
You can find the script here: https://github.com/e-nikolov/snarkOSBlockTest/

It's written in Go, let me know if you are having issues with running it.

**Comment by e-nikolov** [comment]
The difficulty target issue seems to be unrelated. It looks like the numbers are too big and the json parser rounds them. 

```
console.log(JSON.parse(`{"difficulty_target":1844674407370955161}`))
```
The above snippet prints 

```
{
  "difficulty_target": 1844674407370955300
} 
```

Example in the typescript playground:

https://www.typescriptlang.org/play?#code/MYewdgziA2CmB00QHMAUApAygeQHLwAcBDAJwllQAMBvAIgBMBLAM2ceAFdoAXATwH1upZLG60AXAEYAHABZZANgDs8gAxKAzEtUBOAKx7JCyQF9KASnNA

**Comment by ljedrz** [comment]
Thanks, this greatly reduces the scope/severity of the issue.

**Comment by e-nikolov** [comment]
I collected some logs from previous runs of the script:

https://gist.github.com/e-nikolov/8752fa0f5cc69a3140e1a009ebf78bac

**Comment by ljedrz** [comment]
I've looked at the logs and they all seem to be related to block `9b395f3efb734c5c4a7f0145be92a0f717cc37e24d4185997de0bde5a27e4`; if I understand correctly this has happened with some other blocks too, though?

**Comment by e-nikolov** [comment]
I only have the logs from the script for that block.
I'll include the logs from the explorer itself, which are in a different format but show other blocks as well: https://gist.github.com/e-nikolov/deead9539d9d61e7ee5d879a891ec86e

**Comment by ljedrz** [comment]
Thanks to the script, I was able to reproduce the issue locally :fire:; going to try without a secondary storage instance now.

**Comment by ljedrz** [comment]
Can't reproduce with the aforementioned branch. This means that:
- it's not an inconsistency in the primary storage
- we might want to revisit the idea of not using a secondary storage instance at all (it's possible that any past performance issues have been fixed by now)

**Comment by e-nikolov** [comment]
Not sure if it is still relevant, but I modified the script to pick a random block hash from a list and to keep running after it detects the issue rather than stop in order to check how long the issue lasts. I concurrently let 3 versions of the script to run for a couple of hours with 10, 100 and 500 concurrent requests. The same anomalies showed in all logs (except the 500 run missed one because it started a few minutes later) and all lasted for less than 2 seconds. I did not find any patterns in the times when the issues occur - sometimes it was a couple of times within 5 minutes and in other times it did not occur for a long time. 

The logs can be found here: https://gist.github.com/e-nikolov/afcf035f81729beb68e9f5d1c9974cf7
